src/bootstrap/bootstrap.py

```diff
@@ -587,6 +596,61 @@ def fix_bin_or_dylib(self, fname):
+        # If `download-stage1` is not set, default to rebuilding.
+        if self.get_toml("download-stage1", section="rust") != "true":
+            return None
+        # Look for a version to compare to based on the current commit.
+        # There are a few different cases to handle.
+        # 1. This commit is a fast-forward from master: `master - * - * - HEAD`
+        # 2. This commit and master have diverged:
+        # ```
+        #   Y - * - HEAD
+        #  /
+        # X - * - master
+        # ```
+        # In this case, we should compare to `X`.
+        # 3. `master` and `HEAD` are radically different (>100 commits, or similar). This probably
+        # means that `master` does *not* correspond to the version we want to compare to, e.g. a
+        # fork. Instead, we want to compare to `rust-lang/rust:master`, which this has to share a
+        # recent merge base with.
+
+        # Find which remote corresponds to `rust-lang/rust`.
+        remotes = subprocess.check_output(["git", "remote", "-v"], universal_newlines=True)
+        # e.g. `origin https://github.com//rust-lang/rust (fetch)`
+        rust_lang_remote = next(line for line in remotes.splitlines() if "rust-lang/rust" in line)
+        rust_lang_remote = rust_lang_remote.split()[0]
+
+        # Find which commit to compare to
+        merge_base = ["git", "merge-base", "HEAD", "{}/master".format(rust_lang_remote)]
+        commit = subprocess.check_output(merge_base, universal_newlines=True).strip()
+
+        # Next, check if there were changes to the compiler since the ancestor commit.
+        # If so, rebuild instead of using an outdated version of the code.
+        rev_parse = ["git", "rev-parse", "--show-toplevel"]
+        top_level = subprocess.check_output(rev_parse, universal_newlines=True).strip()
+        compiler = "{}/compiler/".format(top_level)
+        # This is slightly more coarse than it needs to be; it rebuild on *any* changes, not just
+        # changes to code.
+        status = subprocess.call(["git", "diff-index", "--quiet", commit, "--", compiler])
+        if status != 0:
+            if self.verbose:
+                print("changes to compiler/ detected, not downloading stage1")
+            return None
+        # Check for untracked files. Adding `build.rs` can change the build
+        # inputs even if no modified files were changed.
+        # Annoyingly, I can't find a way to check for untracked and modified files at the same
+        # time. Even more annoyingly, there doesn't seem to be a way to show *whether* there are
+        # untracked files without listing them.
+        ls_files = ["git", "ls-files", "-o", "--exclude-standard", "compiler/"]
+        untracked_files = subprocess.check_output(ls_files, universal_newlines=True)
+        if untracked_files != '':
+            if self.verbose:
+                print("untracked files in compiler/ detected, not downloading stage1")
+            return None
+        return commit
```

Comment:
I would not try to overcomplicate this. If you've enabled this mode, we should just use the CI artifact regardless of local changes. If it turns out that people get confused by that we can start issuing a warning or error, but I wouldn't try to automatically change behavior for now - it seems much worse to kick off a rustc build *on accident* than to toggle a config file if you realize you do need rustc modifications after all.

Reply:
Yeah, that makes sense - I changed this to warn if `compiler/` was modified and removed the untracked files check altogether.

Reply:
I would prefer to not try to detect modifications at all, personally, it feels like it's not the right approach to get this kicked off - we should land something working and as small as possible first.

Reply:
This is really simple to add though :/ it's just
```python
        # Warn if there were changes to the compiler since the ancestor commit.
        status = subprocess.call(["git", "diff-index", "--quiet", commit, "--", compiler])
        if status != 0:
            print("warning: `download_stage1` is enabled, but there are changes to compiler/")
```